### PR TITLE
[fuzz] Health check fuzz draft

### PR DESCRIPTION
**Pull request by zasweq**
Commit Message: N/A
Additional Description: A draft pr to get comments on initial stages of fuzzer
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

Signed-off-by: Zach Reyes zasweq@google.com

### CODE REVIEW

File: test/common/upstream/health_check_fuzz.cc [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    test::fuzz::Headers headers, absl::string_view status,
+    bool respond_on_second_host) {
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(headers, {}, {}));
+  
+  // If Fuzzer created a status header - replace it with validated status
+  response_headers->setStatus(status);
+
+  int index = (second_host_ && respond_on_second_host) ? 1 : 0;
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.cc:54]: nit: const

File: test/common/upstream/health_check_fuzz.cc [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+  test_sessions_[index]->stream_response_callbacks_->decodeHeaders(
+        std::move(response_headers), true);
+
+
+  //Asserts correct response based on logic in the codebase
+  uint64_t response_code = std::stoi(static_cast<std::string>(status));
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.cc:61]: nit: const

File: test/common/upstream/health_check_fuzz.proto [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+package test.common.upstream;
+
+import "validate/validate.proto";
+import "test/fuzz/common.proto";
+import "envoy/config/core/v3/health_check.proto";
+
+message Respond {
+  test.fuzz.Headers headers = 1;
+  string status =
+      2 [(validate.rules).string.pattern = "([0-9] | [ ])*"];
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.proto:12]: It probably doesn't matter much, but I thought that the codec makes sure that status is a number.

File: test/common/upstream/health_check_fuzz.cc [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  if (response_headers->EnvoyDegraded() != nullptr) {
+    ENVOY_LOG_MISC(trace, "Replied that the host is degraded");
+    ASSERT(cluster_->prioritySet().getMockHostSet(0)->hosts_[index]->health() == Host::Health::Degraded);
+  }
+
+  //No clauses that represent the host not being healthy
+  ENVOY_LOG_MISC(trace, "Replied that the host is Healthy");
+  ASSERT(cluster_->prioritySet().getMockHostSet(0)->hosts_[index]->health() == Host::Health::Healthy);
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.cc:74]: I think the assertion here will fail if the assertion in the previous if statements succeed.
Should all of these be in an "if ... else if ... else ..." clauses?

File: test/common/upstream/health_check_fuzz.cc [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+void HealthCheckFuzz::streamCreate(bool create_stream_on_second_host) {
+  ENVOY_LOG_MISC(trace, "Created a new stream on host 1.");
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.cc:78]: Should probably be either 0 or 1 depending on second_host and create_stream_on_second_host.

File: test/common/upstream/health_check_fuzz.cc [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  if (second_host_) {
+    int index = (create_stream_on_second_host) ? 1 : 0;
+    expectStreamCreate(index);
+    test_sessions_[index]->interval_timer_->invokeCallback();
+  } else {
+    expectStreamCreate(0);
+    test_sessions_[0]->interval_timer_->invokeCallback();
+  }
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.cc:86]: Can probably make it more readable by using something similar to what you did in the above method (`int index = (second_host_ && respond_on_second_host) ? 1 : 0;` and use that for both streams).

File: test/common/upstream/health_check_fuzz.cc [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      break;
+    }
+    case test::common::upstream::HttpAction::kAdvanceTime: {
+        time_system_.advanceTimeAsync(
+            std::chrono::milliseconds(event.advance_time().ms_advanced())
+        );
+        dispatcher_.run(Event::Dispatcher::RunType::NonBlock);
+        break;
+    }
+    default: {
```
Comment (adisuissa) [inline test/common/upstream/health_check_fuzz.cc:112]: nit: brackets are not needed

File: test/fuzz/common.proto [hunk 8]
```diff
@@ -13,7 +13,7 @@ import "validate/validate.proto";
 // Common fuzzing input types.
 
 message Headers {
-  repeated envoy.config.core.v3.HeaderValue headers = 1;
+  repeated envoy.config.core.v3.HeaderValue headers = 1; //TODO: Add headers here
```
Comment (adisuissa) [inline test/fuzz/common.proto:16]: What headers are you expecting here?

**Review by adisuissa — COMMENTED** [review]
Thanks for doing this!
Generally looks good, but one thing I didn't quite understand:
Can you shed some light on why you changed `health_checker_impl_test` to `health_checker_impl_test_utils`?
I think that the unit tests filename should stay as it were before, and if you need some of the functionality, then only that should be refactored into the utils files.

**Comment by asraa** [comment]
This is moving to https://github.com/envoyproxy/envoy/pull/12958?

File: test/common/upstream/health_check_fuzz.proto (hunk 3 above)
Comment (zasweq) [reply test/common/upstream/health_check_fuzz.proto:12]: https://github.com/nodejs/http-parser/blob/d0f3b0548d41acd1d3992f6c36b192620500be58/http_parser.c#L852 Asra linked me this - seems like empty characters are handled?

**Pull request closed without merging**